flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java

```diff
@@ -0,0 +1,462 @@
+			}
+		}
+
+		slotPool.addSlots(acceptedSlots, currentTime);
+		increaseAvailableResources(acceptedResources);
+
+		if (!acceptedSlots.isEmpty()) {
+			notifyNewSlots.accept(acceptedSlots);
+		}
+
+		return acceptedSlotOffers;
+	}
+
+	private Collection<SlotOfferMatching> matchOffersWithOutstandingRequirements(Collection<SlotOffer> slotOffers) {
+		ResourceCounter unfulfilledResources = calculateUnfulfilledResources();
+
+		final Collection<SlotOfferMatching> matching = new ArrayList<>();
+
+		for (SlotOffer slotOffer : slotOffers) {
+			ResourceProfile matchingResourceProfile = null;
+
+			if (unfulfilledResources.containsResource(slotOffer.getResourceProfile())) {
+				unfulfilledResources = unfulfilledResources.subtract(slotOffer.getResourceProfile(), 1);
+
+				// use the profile of the slot offer since it is a direct match against the requirement
+				matchingResourceProfile = slotOffer.getResourceProfile();
+			} else {
+				for (ResourceProfile unfulfilledResource : unfulfilledResources.getResources()) {
+					if (slotOffer.getResourceProfile().isMatching(unfulfilledResource)) {
+						matchingResourceProfile = unfulfilledResource;
+						break;
+					}
+				}
+			}
+
+			final SlotOfferMatching slotOfferMatching;
+			if (matchingResourceProfile != null) {
+				slotOfferMatching = SlotOfferMatching.createMatching(slotOffer, matchingResourceProfile);
+			} else {
+				slotOfferMatching = SlotOfferMatching.createMismatch(slotOffer);
+			}
+
+			matching.add(slotOfferMatching);
+		}
+
+		return matching;
+	}
+
+	@VisibleForTesting
+	ResourceCounter calculateUnfulfilledResources() {
+		return totalResourceRequirements.subtract(fulfilledResourceRequirements);
+	}
+
+	private AllocatedSlot createAllocatedSlot(
+		SlotOffer slotOffer,
+		TaskManagerLocation taskManagerLocation,
+		TaskManagerGateway taskManagerGateway) {
+		return new AllocatedSlot(
+			slotOffer.getAllocationId(),
+			taskManagerLocation,
+			slotOffer.getSlotIndex(),
+			slotOffer.getResourceProfile(),
+			taskManagerGateway);
+	}
+
+	private void increaseAvailableResources(ResourceCounter acceptedResources) {
+		fulfilledResourceRequirements = fulfilledResourceRequirements.add(acceptedResources);
+	}
+
+	@Nonnull
+	private ResourceProfile getMatchingResourceProfile(AllocationID allocationId) {
+		return Preconditions.checkNotNull(slotToRequirementProfileMappings.get(allocationId), "No matching resource profile found for %s", allocationId);
+	}
+
+	@Override
+	public PhysicalSlot reserveFreeSlot(AllocationID allocationId, ResourceProfile requiredSlotProfile) {
+		final AllocatedSlot allocatedSlot = slotPool.reserveFreeSlot(allocationId);
+
+		Preconditions.checkState(allocatedSlot.getResourceProfile().isMatching(requiredSlotProfile), "");
+
+		ResourceProfile previouslyMatchedResourceProfile = Preconditions.checkNotNull(slotToRequirementProfileMappings.get(allocationId));
+
+		updateSlotToRequirementProfileMapping(allocationId, requiredSlotProfile);
```

Comment:
I'm not sure if we can get _rid_ of it.

The information that _this_ slot was used to fulfill _that_ requirement is used for so many things; be it matching slot offers against outstanding requirements, figuring out whether a slot is eligible for an idleness timeout, or to update the fulfilled requirements if a slot is released.
It seems impossible to remove this mapping without removing everything that makes this pool more than a wrapper around the `AllocatedSlotPool`.

At the end of the day we need this information _somewhere_ anyway; be it in the slot pool, or the new scheduler so it could make sure to update the requirements.
Given that the declarative slot pool is all about managing slots in respects to requirements, it seems fitting that it contains such a mapping.

Maybe this is just a naming/structural issue? What if the DeclarativeSlotPool were not actually a pool, but a requirement manager?
Slot offers and releases would be issued directly against he AllocatedSlotPool, which informs the declarative slot pool, which then takes the appropriate steps of adjusting requirements etc. .
Basically, instead of removing the mapping portion out of the slot pool, remove the slot pool portion from it.

Reply:
I think my main concern comes from the following thought: I've considered the `DeclarativeSlotPool` as the component which makes sure that we keep a set of slots according to the declared resources and makes it accessible to a user for usage. If now the user decides to use a slot differently (e.g. for a smaller slot request), then it should be his responsibility to make sure that he no longer needs this slot or to adjust the resource requirements by asking for another big slot in order to fulfill the pending slot requests. Since this is something which has to be decided by the user, it shouldn't bother this component.

What I am not saying is that we cannot remember the mapping of a matching to speed up future matchings. This is perfectly fine and can be encapsulated in the implementation. So what bothers me is that the `reserveFreeSlot` method has a `requiredSlotProfile ` argument.